[Yang Yuzhao] iP

Morrow1ndy opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen
> ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/Morrow1ndy/ip/releases/tag/v1.0).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. 
Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
      ......
   }
}
```

Commit:
Add Gradle support

Commit:
Level 1

Commit:
Level 2

Commit:
Level 3

Commit:
Level 4

Commit:
A-TextUiTesting

Commit:
Level 5

Commit:
Modified Level 5

Minor change.

Commit:
Level 5.

Bug fix.

Commit:
Level 6 and A-Enums.

Finished everything(minimal requirement).

Commit:
Implement Level 6 and A-Enums.

Restructure the project by moving every class to an appropriate package.

Commit:
Complete Level 7.

Save is available in Duke.

Commit:
Complete Level 8.

Dates and Times are supported in Duke.

Commit:
Revert "Complete Level 8."

This reverts commit 79023c5ac0348631c6573661b62477f1d58cdd4b.

Commit:
Minor Clean-up.

Commit:
Complete Level 7.

Minor Clean-up.
Save is supported in Duke.

Commit:
Solve Merge Confilct.

Commit:
Redundant file clean-up.

Commit:
Merge branch 'branch-Level-8'

Commit:
Complete Level 8.

Dates and Times are supported in Duke.

Commit:
Complete A-MoreOOP.

Classes are extracted.

Commit:
Complete A-Packages.

Commit:
Complete A-Junit.

Commit:
Complete A-Jar.

Commit:
Complete A-JavaDoc.

Commit:
Complete A-CodingStandard.

Commit:
Complete Level-9.

Find is supported in Duke.

Minor bug:  text-ui-test does not work properly

Commit:
Minor change.

eeliana commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public class Duke {
+
+	public static void main(String[] args) {
+		Ui.showWelcomeMessage();
```
Should indentation be 4 spaces instead?

Morrow1ndy replied:
Thanks for reminding me!

Morrow1ndy replied:
The indentation shown on my IDE is 4 spaces on my IDE. I'm not sure whether it got something to do with git.

eeliana commented on src/main/java/duke/operation/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+	@Override
+	public String toString() {
+		DateTimeFormatter dateTimeFormatterTo = DateTimeFormatter.ofPattern("MMM-dd-yyyy HH:mm");
+		return "[D]"
+				+ super.toString()
+				+ " (by: "
+				+ this.by.format(dateTimeFormatterTo)
+				+ ")";
```
I like the way you format this such that it is very clean and easy to read but I am not sure if this fits the module's coding standards :0

Morrow1ndy replied:
Thanks for reminding and I'll take a look at the coding standards again:)

Morrow1ndy replied:
The code format is shown differently on GitHub vs IDE. Not sure why this happens.

eeliana commented on src/test/java/duke/operation/DeadlineTest.java:
```diff
@@ -0,0 +1,19 @@
+import java.time.format.DateTimeFormatter;
+
+public class DeadlineTest {
+	private final DateTimeFormatter dateTimeFormatterFrom = DateTimeFormatter.ofPattern("yyyy-dd-MM HH:mm");
```
Is this more than 120chars on the line?

Morrow1ndy replied:
Thanks! Will re-format it.

eeliana reviewed (commented):
LGTM after a few coding standard violations to correct. Overall, the code was clean and very easy to read 👍

SajalVaishnav commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+		storage.createFile();
+		TaskList taskList = new TaskList();
+
+		while (true) {
```
Maybe you can separate out this logic into a factory class?

Morrow1ndy replied:
Yes I can. Maybe I'll do it some time later.

SajalVaishnav commented on src/main/java/duke/Parser/Parser.java:
```diff
@@ -0,0 +1,50 @@
+	 * @throws DukeException
+	 */
+	public static void invalidTask() throws DukeException {
+		String message = "____________________________________________________________\n"
```
I really like how you've kept the layout neat with line separators!

Morrow1ndy replied:
Thanks!

SajalVaishnav commented on src/main/java/duke/operation/TaskList.java:
```diff
@@ -0,0 +1,214 @@
+/**
+ * This is the TaskList class to contain task list and handle operations.
+ */
+public class TaskList {
```
This is a huge class, simultaneously maintaining the list of tasks and handling the I/O too... would it be possible to maybe delegate some of its responsibilities?

SajalVaishnav commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * This is the DukeException class to handle Duke-specific exceptions.
+ */
+public class DukeException extends Exception {
```
Maybe one or two more exception types could help you handle a bigger range of exception-al scenarios?

Morrow1ndy replied:
I have thought of it before. Since current implementation already satisfy the minimum requirement, maybe I can do it when I am free.

Commit:
Re-format the code.

In order to meet the module's coding standards.

SajalVaishnav reviewed (commented):
LGTM! minor nits

Commit:
Use GFMD in README.

Commit:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

Commit:
Fix Gradle configuration bug.

Commit:
Revert "Fix Gradle configuration bug."

This reverts commit 69bf03dc488a27031df454e3b7de32b452a0fd3c.

Commit:
Add Gradle support.

Run the project using Gradle.

angkoonhwee commented on src/main/java/duke/operation/Command.java:
```diff
@@ -0,0 +1,64 @@
+ * This is Command class to contain command strings to separate
+ * corresponding operations accordingly.
+ */
+public enum Command {
```
good method of avoiding "magic strings"!

angkoonhwee commented on src/main/java/duke/operation/TaskList.java:
```diff
@@ -0,0 +1,214 @@
+	 * @param inputLine input string
+	 * @throws DukeException
+	 */
+	public void addTask(Command command, String inputLine) throws DukeException {
```
Perhaps you can extract out certain sections of the code such as the constants for printAddTask to a variable outside the method since the method is too long and according to the coding standards here methods should not be longer than 30 LOC [here](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods)

angkoonhwee commented on src/main/java/output/duke.txt:
```diff
@@ -0,0 +1,2 @@
+[T][ ] 111
```
perhaps you can include txt files in `.gitignore` files so that your storage file wouldn't be uploaded to public

angkoonhwee reviewed (commented):
Actually looks good to me, my comments are just being nitpicky! 👍

Commit:
Update Duke GUI without Duke logic.

Profile photos to be added later on.

Commit:
Basic GUI for Duke.

Commit:
Add profile pictures to GUI.

Profile pictures are needed to show the sender of the messages in GUI.

Different profile pictures are added to 2 parties: Duke and user.

Commit:
Add assertion to all types of tasks.

Assertions are to ensure the assumptions of the input format is meet.

Assertions of certain keywords and data format is checked.

Morrow1ndy committed:
Merge pull request #1 from Morrow1ndy/A-Assertions

Add assertion to all types of tasks.

Commit:
Improve code quality.

Code is improved according to coding convention.

Morrow1ndy committed:
Merge pull request #2 from Morrow1ndy/A-CodeQuality

Improve code quality.

Commit:
Add auto sorting by task type.

When list command is input, output task list should be sorted.

Task list is auto sorted by the ascending order: todo, deadline, event
with "list" command.

Commit:
Fix dialog height bug in GUI.

DialogBox's height should be flexible.

A dynamic dialogBox height can include as many messages as possible that
are returned.

Let's:
	- update the label property in DialogBox fxml file

Commit:
Update sorting to deadline and event tasks.

Deadline and events should be auto-sorted based on their date/deadlines.

Deadlines and events are sorted by an ascending order of
their corresponding dates.

Morrow1ndy committed:
Set theme jekyll-theme-architect

Commit:
Update README.

README should contain User Guide.

Commit:
Merge branch 'master' of https://github.com/Morrow1ndy/ip

Commit:
Add Ui preview.

Commit:
Fix a bug that bye command does not exit the application correctly.

Commit:
Add abstraction to fit OOP requirements and fix minor bugs.
